src/main/scala/co/teapot/mmalloc/PieceBlock.scala

```diff
@@ -83,63 +84,66 @@ private[mmalloc] class PieceBlock(val pointer: Long, mmAllocator: MemoryMappedAl
+    require(pieceSize.raw < Int.MaxValue, "Piece size has to fit into Integer type")
+    data.putInt(pointer + PieceSizeOffset, pieceSize.raw.toInt, syncWrite)
   }
 
-  def pieceCount(pointer: Long, data: LargeMappedByteBuffer): Int =
+  def pieceCount(pointer: Pointer, data: LargeMappedByteBuffer): Int =
     data.getInt(pointer + PieceCountOffset)
 
-  def freeCount(pointer: Long, data: LargeMappedByteBuffer): Int =
+  def freeCount(pointer: Pointer, data: LargeMappedByteBuffer): Int =
     data.getInt(pointer + FreeCountOffset)
-  def setFreeCount(pointer: Long, data: LargeMappedByteBuffer, freeCount: Int, syncWrite: Boolean): Unit = {
+  def setFreeCount(pointer: Pointer, data: LargeMappedByteBuffer, freeCount: Int, syncWrite: Boolean): Unit = {
     data.putInt(pointer + FreeCountOffset, freeCount, syncWrite)
   }
 
-  def isFull(pointer: Long, data: LargeMappedByteBuffer): Boolean =
+  def isFull(pointer: Pointer, data: LargeMappedByteBuffer): Boolean =
     freeCount(pointer, data) == 0
 
-  def searchStart(pointer: Long, data: LargeMappedByteBuffer): Int =
+  def searchStart(pointer: Pointer, data: LargeMappedByteBuffer): Int =
     data.getInt(pointer + PieceBlock.SearchStartOffset)
-  def setSearchStart(pointer: Long, data: LargeMappedByteBuffer, searchStart: Int): Unit = {
+  def setSearchStart(pointer: Pointer, data: LargeMappedByteBuffer, searchStart: Int): Unit = {
     // Don't need to sync, since searchStart is just an optimization hint
     data.putInt(pointer + PieceBlock.SearchStartOffset, searchStart)
   }
 
   def initializePieceBlock(
-      pointer: Long,
+      pointer: Pointer,
       data: LargeMappedByteBuffer,
-      pieceSize: Int,
-      blockSize: Int,
+      pieceSize: ByteCount,
+      blockSize: ByteCount,
       syncWrite: Boolean): Unit = {
     setPieceSize(pointer, data, pieceSize, false)
     // For each piece, we store an extra bit (1.0 / 8 bytes) to track if its free.  Hence
     // the number of pieces we can fit is the blockSize (excluding header and 4 bytes for rounding)
     // divided by the pieceSize + 1.0/8
-    val pieceCount = ((blockSize - PieceBlock.BitsetOffset - 4) / (pieceSize + 1.0 / 8)).toInt
+    val pieceCount = ((blockSize.raw - PieceBlock.BitsetOffset.raw - 4) / (pieceSize.raw + 1.0 / 8)).toInt
     data.putLong(pointer + BlockTypeOffset, PieceBlockMagicNumber)
-    data.putInt(pointer + PieceSizeOffset, pieceSize)
+    require(pieceSize.raw < Int.MaxValue)
+    data.putInt(pointer + PieceSizeOffset, pieceSize.raw.toInt)
     data.putInt(pointer + PieceCountOffset, pieceCount)
     data.putInt(pointer + FreeCountOffset, pieceCount)
     data.putInt(pointer + SearchStartOffset, 0)
     ByteBufferBasedBitSet.initializeWith1s(pointer + BitsetOffset, data, pieceCount)
     if (syncWrite) {
-      data.syncToDisk(pointer, BitsetOffset + pieceCount / 8)
+      data.syncToDisk(pointer, ByteCount(BitsetOffset.raw + pieceCount / 8))
```

Comment:
You don't have to, but this might be clearer as something like
`ByteCount(BitsetOffset + Offset.bits(pieceCount))`